Some improvements based on the use of grub-btrfs

kevinchevreuil opened the pull request:
Hello the team,

I hope you are well.

When using grub-btrfs and the integration I did previously on Kaisen Linux, caused me some important bugs (detailed below).

I have reviewed my integration and offer you some code improvements:

- Removal of grub-btrfs.cfg also when grub-btrfs.new is deleted (if no snapshot exists, why keep the file? Moreover, even if all snapshots are deleted, the file still keeps some snapshots that can't be booted, thus creating false entries, this scenario works on Debian (for my part tested on Kaisen Linux 2.2) when only the script /etc/grub.d/41_snapshots-btrfs is executed and not the command grub-mkconfig -o /boot/grub/grub.cfg). Commit: https://github.com/kevinchevreuil/grub-btrfs/commit/9118c43c598666c35e36aab83d095c3523d8c65b

- Creation of an if statement to no longer display the BTRFS snapshots menu if the file /boot/grub/grub-btrfs.cfg does not exist (if no snapshot exists, it will not be created, in connection with the first proposal). This menu does disappear when the grub-mkconfig -o /boot/grub/grub.cfg command is executed, but not the /etc/grub.d/41_snapshots-btrfs script alone, causing an invalid entry. In connection with the previous commit, it fixes this problem, and allows to execute only the script alone without creating an invalid GRUB entry. Commit: https://github.com/kevinchevreuil/grub-btrfs/commit/a705b06cff758db788ce02d5d690a4a322135e90

- The grub-btrfsd script on the create_grub_menu function only executes the /etc/grub.d/41_snapshots-btrfs script in connection with the previous commits. Commit: https://github.com/kevinchevreuil/grub-btrfs/commit/1b6a9d44a19f0a06de2c8aa78f29ea054d2a763c

- I added to the Makefile an if statement to run an update of GRUB when installing and uninstalling grub-btrfs. This allows to add at the installation of grub-btrfs the condition to display BTRFS snapshots in the menu if they exist in connection with the previous commits or to remove them from GRUB if the program is uninstalled. This thing is also done on my default Debian package. I'm trying to enforce this behavior for a standard installation without package. Commits: https://github.com/kevinchevreuil/grub-btrfs/commit/a58b6cbb5a4ed4ad19311fdc7eb00e4b7d5bb43a and https://github.com/kevinchevreuil/grub-btrfs/commit/263cc1ddf9279a1e4231feb60d581db9a884ba57

The reason why I suggest these improvements is related to the integration of grub-btrfs in Kaisen Linux via my Debian package. I was running the grub-mkconfig -o /boot/grub/grub.cfg command and not the standalone script. All this was coupled with a systemd .path and .service unit file (.path checking the /run/timeshift directory), precisely to "fix" this problem. I found the same problems (invalid GRUB entries) when using the grub-btrfsd daemon since 4.12. Except that when using grub-btrfs to restore snapshots, the grub menu was completely updated and changed the GRUB entry of Kaisen by the path of the snapshot... If this one is deleted, the system becomes "impossible" to boot, and especially, it is not the good sub-volume which is used after the restoration of a snapshot...
Indeed, launching only the /etc/grub.d/41_snapshots-btrfs script worked, but left invalid GRUB entries. Since 4.12, if the GRUB update was not done before or the /boot/rub/grub-btrfs.cfg file is empty, the whole grub menu is updated, which can be problematic in some specific use cases, for example if you want to have a fully dynamic GRUB menu without regenerating the whole configuration each time (or to limit the various problems linked to snapshots startups and/or users behaviors).

If these proposals are accepted, how about creating a 4.12.1 or 4.13 release (at your convenience), so that I can redistribute these default fixes in the unpatched code on my Debian package as well as to allow potential other maintainers of grub-btrfs on a distribution to integrate the latest default fixes?

Antynea commented:
hello,

Thanks!

> Removal of grub-btrfs.cfg also when grub-btrfs.new is deleted ...

Indeed, the current implementation is bad.
Perhaps the process could be strengthened by creating a backup of "grub-btrfs.cfg" before it is deleted, so that it can be restored in case of an error.

> Creation of an if statement to no longer display the BTRFS snapshots menu if the file ...
Great, it works well.

> The grub-btrfsd script on the create_grub_menu function only executes the /etc/grub.d/41_snapshots-btrfs

Yes it's a better behaviour than the original, as you mentioned and I completely agree with that, we don't want the Grub menu to be rebuilt when booting in a snapshot.
I hope @Schievel1 will be able to validate the proper functioning.

> I added to the Makefile an if statement to run an update of GRUB when installing and uninstalling grub-btrfs ...

I like the idea, however fedora uses grub2-mkconfig instead of grub-mkconfig, it won't work for them.

Ps: Although this is not part of your pr, I noticed that the grub-btrfs.new file was still present when the script stops when there is no snapshot. My fault, I didn't add an instruction to delete it at that time.

kevinchevreuil commented:
Hello, thank you for your answer Antynea :)

In this case, you can create a .bkp file for example before deleting it.
I will review my PR.

I'll add in my PR the deletion of the grub-btrfs.new file.

For the makefile and the execution of the update of GRUB, I can create an if statement to exec the command associated to the distribution, grub-mkconfig for Debian and based, grub2-mkconfig for Fedora and based...

All combinated, grub-btrfs will be unstoppable ;)

kevinchevreuil commented:
@Antynea I will review my PR, I add some stuff ;)

kevinchevreuil commented:
@Antynea I will review my PR, I add some stuff ;)

kevinchevreuil commented:
@Antynea I have tested all changes. I just push an update of my PR :)

Enjoy :)

kevinchevreuil commented:
@Antynea I have tested all changes. I just push an update of my PR with clean old bad commit :)

I also included a backup of the file rather than deleting it, if all goes well, everything will remain transparent for the user :)

I also added a variable for RHEL users for GRUB in the Makefile, the update works for Fedora too :)

The .new is now deleted if no snapshot has been found (same for the backup file :))

Enjoy :)

Schievel1 commented:
Hey, thanks for your awesome work. I really appreciate it.

> @Antynea I have tested all changes. I just push an update of my PR with clean old bad commit 

Yes, please rebase this a bit and clean it up. I usually prefix the commits with the part I changed, e.g. `grub-btrfsd:` or `README:`

Please give me some time to check those things out first. I am a bit busy doing family stuff these days. :)

kevinchevreuil commented:
Hello @Schievel1 and @Antynea, happy new year! :)

I rebase my latest commits.

No problem for me :) enjoy ;)

Schievel1 commented:
> > The grub-btrfsd script on the create_grub_menu function only executes the /etc/grub.d/41_snapshots-btrfs
> 
> Yes it's a better behaviour than the original, as you mentioned and I completely agree with that, we don't want the Grub menu to be rebuilt when booting in a snapshot. I hope @Schievel1 will be able to validate the proper functioning.

I think we didn't do this ever. We check for existence of grub-btrfs.cfg and run only 41_snaphshots-btrfs if it exists.

kevinchevreuil commented:
Hello @Schievel1,

What is the problem?

In this PR, I purpose to delete the grub-btrfs.cfg file if no snapshot was found.

Schievel1 commented:
Thank you very much.
